forem/forem
# Data Update Script to migrate logo_svg contents to png file (Will be merged and deployed on 10/01)

## What type of PR is this? (check all applicable)


## Description
As per https://forem.team/engineering/creator-signup-and-onboarding-16p6, we've updated the logo upload process for creators as follows:
- We only accept PNG/JPG/JPEG to be uploaded as a logo.
- We no longer accept SVG due to security vulnerabilities.

Based on the above we need to convert existing Forem’s logos from SVG (i.e, the `logo_svg` markup) to PNG so that there is no action for Forem Creators. 

Please note that this is a best effort script and we try to cater for as many conditions/edge cases as possible but we are unable to cater for everything. Hence we will be giving creators the opportunity to change/tweak their logos and re-upload if the resolution, visibility has been altered during the conversion. In addition, we will remind them that they should view their logo on both dark and light theme.

There are some Forems (very few it seems, like DEV, forem.dev and vscodetips) who have added custom scripts or are using dynamic CSS variables in their SVG logo to alter the logo’s styles for our dark theme and light theme (an example is the DEV logo). We are not able to convert these logos accurately with our tools.

Hence, we do our best effort to first provide static values for these CSS variable (that we know will not work in dark theme) and for others we want this script to log the errors for any of the logos that fail to convert.

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/15620

## QA Instructions, Screenshots, Recordings

_Test a successful update:_ 
1. Open a rails console and set the value for `logo_svg` 
```
Settings::General.logo_svg = '<svg width="300" height="100" xmlns="http://www.w3.org/2000/svg">
    <path fill="#dedede" stroke="#555" stroke-width="2" d="M2 2h296v96H2z"/>
    <text x="50%" y="50%" font-size="18" text-anchor="middle" alignment-baseline="middle" font-family="monospace, sans-serif" fill="#555">
        300×100
    </text>
</svg>'
```

2. Make sure that the logo's are nil. Open a rails console and run:
```
Settings::General.original_logo = nil 
Settings::General.resized_logo = nil
```
3. Run the Data Update Script in your terminal with `data_updates:run`
4. Open the rails console and check that the values for `Settings::General.original_logo` and `Settings::General.resized_logo` have updated.
5. Go to `/public/uploads/logos` and you should see an original and resized png versions of the original svg. 


You can also test a logo with a css variable like `currentColor` or `var(--base-inverted)`  and watch it convert if you'd like.
### UI accessibility concerns?

None.

## Added/updated tests?

      have not been included_
- [ ] I need help 

Some insights into fixing the Travis tests:
Initially the MiniMagick conversion process is failing on Travis.
```
CarrierWave::ProcessingError:
       Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /home/travis/build/forem/forem/tmp/1640166684-774651839101422-2549-8379/300x100.svg -auto-orient /tmp/image_processing20211222-11351-73ymxc.png` failed with error:
       convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
       convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
     # ./spec/uploaders/logo_svg_uploader_spec.rb:39:in `block (3 levels) in <main>'
     # ./spec/[secure]_helper.rb:145:in `block (3 levels) in <main>'
     # ./spec/[secure]_helper.rb:145:in `block (2 levels) in <main>'
     # ./spec/[secure]_helper.rb:100:in `block (2 levels) in <main>'
     # ------------------
     # --- Caused by: ---
     # MiniMagick::Error:
     #   `convert /home/travis/build/forem/forem/tmp/1640166684-774651839101422-2549-8379/300x100.svg -auto-orient /tmp/image_processing20211222-11351-73ymxc.png` failed with error:
     #   convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
     #   convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
     #   ./spec/uploaders/logo_svg_uploader_spec.rb:39:in `block (3 levels) in <main>'

```

 I tried a couple of things thus far like:
- Making sure that the fixture file is found as per [Mac's comment](https://github.com/forem/forem/pull/15710#discussion_r773195512)
- Testing the functionality with the fixture file locally (I moved it temporarily into the uploads folder): 
```
svg_file = File.open('./public/uploads/300x100.svg')
logo_svg_uploader = LogoSvgUploader.new.tap do |uploader|
  uploader.store!(svg_file)
end
```
 and it worked.
- https://travis-ci.community/t/build-times-out-with-no-apparent-reason/5083 as per this [commit](https://github.com/forem/forem/pull/15710/commits/53d96d006af2b7b6ab90344a336c294abe54639e).
- https://stackoverflow.com/questions/41138404/how-to-install-newer-imagemagick-with-webp-support-in-travis-ci-container as per this [commit](https://github.com/forem/forem/pull/15710/commits/9835d466e66e0eebfdfcdfc0b889b40acc3b3453)

I suspect it's a Travis configuration issue or maybe version difference in Travis, but haven't been able to figure it out as yet and would appreciate any insights on it.

See more of the conversation here: https://forem-team.slack.com/archives/C7GE84DEJ/p1641305439053700

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

I've chatted to Jennie about alerting Forems about the following: 
Forems will have a week to preview their converted uploads to check for:
1. resolution 
2. visibility 
3. and determine if they'd like to twek/update their logo to work on both dark adn light theme

Pawel will be tweaking our logos as well. 

We will send out a forem.dev post once we have run the script.

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)
